src/test/run-pass/issue-30490.rs

```diff
@@ -0,0 +1,90 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Previously libstd would set stdio descriptors of a child process
+// by `dup`ing the requested descriptors to inherit directly into the
+// stdio descriptors. This, however, would incorrectly handle cases
+// where the descriptors to inherit were already stdio descriptors.
+// This test checks to avoid that regression.
+
+#![cfg_attr(unix, feature(libc))]
+
+use std::fs::{self, File};
+use std::io::{Read, Write};
+use std::io::{stdout, stderr};
+use std::process::{Command, Stdio};
+
+#[cfg(unix)]
+fn main() {
+    extern crate libc;
+
+    use std::os::unix::io::FromRawFd;
+    use std::os::unix::io::IntoRawFd;
+
+    let stdout_backup = unsafe { libc::dup(libc::STDOUT_FILENO) };
+    let stderr_backup = unsafe { libc::dup(libc::STDERR_FILENO) };
+
+    assert!(stdout_backup > -1);
+    assert!(stderr_backup > -1);
+
+    let stdout_file_path = concat!("test-", module_path!(), "-out");
+    let stderr_file_path = concat!("test-", module_path!(), "-err");
+
+    {
+        let stdout = File::create(&stdout_file_path).expect("failed to create stdout file");
+        let stderr = File::create(&stderr_file_path).expect("failed to create stderr file");
```

Comment:
Could this test be written without creating files? Keeping everything in-memory is generally better for isolation. I think this should basically just be able to use combinations of pipes, right?